feed/feed.go

```diff
@@ -0,0 +1,69 @@
+package feed
+
+import (
+	"sync"
+)
+
+type Feed[T any] struct {
+	mu     sync.Mutex // protects nextID and subs.
+	subs   map[uint64]*Subscription[T]
+	nextID uint64
+}
+
+type Subscription[T any] struct {
+	c         chan T
+	f         *Feed[T]
+	unsubOnce sync.Once
+	id        uint64
+}
+
+func (s *Subscription[T]) Recv() <-chan T {
+	return s.c
+}
+
+func (s *Subscription[T]) Unsubscribe() {
+	s.unsubOnce.Do(func() {
+		s.f.mu.Lock()
+		defer s.f.mu.Unlock()
+		for _, sub := range s.f.subs {
+			if sub.id == s.id {
+				close(sub.c)
+				delete(s.f.subs, sub.id)
+				return
+			}
+		}
+	})
+}
+
+func New[T any]() *Feed[T] {
+	return &Feed[T]{
+		subs: make(map[uint64]*Subscription[T], 0),
+	}
+}
+
+func (f *Feed[T]) Subscribe() *Subscription[T] {
+	ch := make(chan T, 1)
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	s := &Subscription[T]{
+		c:  ch,
+		f:  f,
+		id: f.nextID,
+	}
+	f.nextID++
```

Comment:
Let's use [atomics](https://pkg.go.dev/sync/atomic#Uint64.Add) to be safe here

Reply:
Done, although `f.mu` should be sufficient to protect `nextID`.

Reply:
Ah, my bad. Didn't realize that.